README.rst
=====
term
=====
--------------------------------------
An enhanced version of the tty module
--------------------------------------

Overview
========

The **term** package is an enhanced version of the standard library's
tty_ module. It provides context managers for opening a terminal stream, and
for temporarily switching the terminal to raw or cbreak mode.

.. _tty: https://docs.python.org/3/library/tty.html

Package Contents
================

Terminal Control
----------------

setraw(fd, when=TCSAFLUSH, min=1, time=0)
    Put the terminal in raw mode.

setcbreak(fd, when=TCSAFLUSH, min=1, time=0)
    Put the terminal in cbreak mode.

rawmode(fd, when=TCSAFLUSH, min=1, time=0)
    Context manager to put the terminal in raw mode.

cbreakmode(fd, when=TCSAFLUSH, min=1, time=0)
    Context manager to put the terminal in cbreak mode.

Terminal I/O
------------

opentty(bufsize=-1, mode='r+b')
    Context manager returning a new rw stream connected to /dev/tty.
    The stream is None if the device cannot be opened.

readto(stream, endswith):
    Read bytes or characters from stream until buffer.endswith(endswith)
    is true.

High-level Functions
--------------------

getyx()
    Return the cursor position as 1-based (line, col) tuple.
    Line and col are 0 if the device cannot be opened or the terminal
    does not support DSR 6.

getfgcolor()
    Return the terminal foregound color as (r, g, b) tuple.
    All values are -1 if the device cannot be opened or does not supports
    OSC 10.

getbgcolor()
    Return the terminal background color as (r, g, b) tuple.
    All values are -1 if the device cannot be opened or does not supports
    OSC 11.

islightmode()
    Return true if the background color is lighter than the foreground color.
    May return None if the terminal does not support OSC color queries.

isdarkmode()
    Return true if the background color is darker than the foreground color.
    May return None if the terminal does not support OSC color queries.

Documentation
=============

Please see the `API Documentation`_ for more.

.. _`API Documentation`: https://term.readthedocs.io/en/stable/



CHANGES.rst
Changelog
=========

2.6 - Unreleased
----------------


2.5 - 2023-09-14
----------------

- Update tox.ini for latest tox.
  [stefan]

- Add .readthedocs.yaml file.
  [stefan]

- Pin sphinx and sphinx-rtd-theme versions in docs extra.
  [stefan]

- Add pylint extra which installs pylint.
  [stefan]

- Update .pylintrc for latest pylint.
  [stefan]

2.4 - 2022-03-07
----------------

- Add Python 3.8-3.10 to tox.ini. Remove old Python versions.
  [stefan]

- Replace deprecated ``python setup.py test`` in tox.ini.
  [stefan]

- Remove deprecated ``test_suite`` from setup.py.
  [stefan]

- Move metadata to setup.cfg and add a pyproject.toml file.
  [stefan]

- Include tests in sdist but not in wheel.
  [stefan]

- Fix escape sequence warning in byte string literal.
  [stefan]

- Open /dev/tty in binary mode under both Python 2 and 3.
  [stefan]

- Officially change opentty's bufsize argument default from 1 to -1.
  Under Python 3, 1 has effectively meant -1 all along but Python 3.8
  now issues a warning.
  [stefan]

- Add readto, getfgcolor, getbgcolor, islightmode, and isdarkmode.
  [stefan]

2.3 - 2019-02-08
----------------

- Add MANIFEST.in.
  [stefan]

- Release as wheel.
  [stefan]

- Drop explicit GPL because the PSF license is GPL-compatible anyway.
  [stefan]

2.2 - 2017-02-05
----------------

- Support Python 2.6-3.6 without 2to3.
  [stefan]

2.1 - 2014-04-19
----------------

- Remove setuptools from install_requires because it isn't.
  [stefan]

2.0 - 2012-04-27
----------------

- Open /dev/tty in binary mode under Python 3.
  [stefan]

- Disable buffering if the device is not seekable.
  [stefan]

- Remove getmaxyx since it cannot be implemented reliably.
  [stefan]

- Support Python 2.5.
  [stefan]

- Change license to GPL or PSF to avoid relicensing of PSF code.
  [stefan]

1.0 - 2012-04-11
----------------

- Initial release.
  [stefan]


term/tests/test_term.py
import sys
import os
import unittest
import termios

# pylint: disable=wildcard-import
# pylint: disable=unused-wildcard-import
from termios import *
from term import *

from term import MODE
from term import _opentty
from term import _readyx
from term import _readcolor
#from term import getyx
from term import getbgcolor
from term import getfgcolor
from term import luminance
from term import islightmode
from term import isdarkmode
from term import isxterm

if sys.version_info[0] >= 3:
    from io import BytesIO
    from io import StringIO
else:
    from StringIO import StringIO as BytesIO
    from StringIO import StringIO


# CC members VMIN and VTIME should be ints, not bytes.
# CPython and PyPy return ints in some cases and bytes in others.
def int_(x):
    if not isinstance(x, int):
        x = ord(x)
    return x


class setterm(object):
    def __init__(self, val):
        self._val = val
        self._saved = None
    def __enter__(self):
        self._saved = os.environ.get('TERM', '')
        os.environ['TERM'] = self._val
    def __exit__(self, *ignored):
        os.environ['TERM'] = self._saved


class IntegerTests(unittest.TestCase):

    def test_int(self):
        self.assertEqual(int_(0), 0)
        self.assertEqual(int_(1), 1)
        self.assertEqual(int_(255), 255)
        self.assertEqual(int_(4), 4)
        self.assertEqual(int_(32767), 32767)

    def test_bytes(self):
        self.assertEqual(int_(b'\x00'), 0)
        self.assertEqual(int_(b'\x01'), 1)
        self.assertEqual(int_(b'\xff'), 255)
        self.assertEqual(int_(b'4'), 52)
        self.assertEqual(int_(b'\x34'), 52)

    def test_bytes_too_long(self):
        self.assertRaises(TypeError, int_, b'32767')


class TermTests(unittest.TestCase):
    # pylint: disable=too-many-public-methods

    def setUp(self):
        self.savedmode = tcgetattr(sys.stdin)

    def tearDown(self):
        tcsetattr(sys.stdin, TCSAFLUSH, self.savedmode)

    def test_defaults(self):
        mode = tcgetattr(sys.stdin)
        self.assertEqual(mode[LFLAG] & ECHO, ECHO)
        self.assertEqual(mode[LFLAG] & ICANON, ICANON)
        self.assertEqual(mode[LFLAG] & IEXTEN, IEXTEN)
        self.assertEqual(mode[LFLAG] & ISIG, ISIG)
        self.assertEqual(int_(mode[CC][VMIN]), 1)
        self.assertEqual(int_(mode[CC][VTIME]), 0)

    def test_setraw(self):
        setraw(sys.stdin, min=0, time=1)
        mode = tcgetattr(sys.stdin)
        self.assertEqual(mode[LFLAG] & ECHO, 0)
        self.assertEqual(mode[LFLAG] & ICANON, 0)
        self.assertEqual(mode[LFLAG] & IEXTEN, 0)
        self.assertEqual(mode[LFLAG] & ISIG, 0)
        self.assertEqual(int_(mode[CC][VMIN]), 0)
        self.assertEqual(int_(mode[CC][VTIME]), 1)

    def test_setcbreak(self):
        setcbreak(sys.stdin, min=0, time=1)
        mode = tcgetattr(sys.stdin)
        self.assertEqual(mode[LFLAG] & ECHO, 0)
        self.assertEqual(mode[LFLAG] & ICANON, 0)
        self.assertEqual(mode[LFLAG] & IEXTEN, IEXTEN)
        self.assertEqual(mode[LFLAG] & ISIG, ISIG)
        self.assertEqual(int_(mode[CC][VMIN]), 0)
        self.assertEqual(int_(mode[CC][VTIME]), 1)

    def test_rawmode(self):
        with rawmode(sys.stdin, min=0, time=1):
            mode = tcgetattr(sys.stdin)
            self.assertEqual(mode[LFLAG] & ECHO, 0)
            self.assertEqual(mode[LFLAG] & ICANON, 0)
            self.assertEqual(mode[LFLAG] & IEXTEN, 0)
            self.assertEqual(mode[LFLAG] & ISIG, 0)
            self.assertEqual(int_(mode[CC][VMIN]), 0)
            self.assertEqual(int_(mode[CC][VTIME]), 1)
        self.test_defaults()

    def test_cbreakmode(self):
        with cbreakmode(sys.stdin, min=0, time=1):
            mode = tcgetattr(sys.stdin)
            self.assertEqual(mode[LFLAG] & ECHO, 0)
            self.assertEqual(mode[LFLAG] & ICANON, 0)
            self.assertEqual(mode[LFLAG] & IEXTEN, IEXTEN)
            self.assertEqual(mode[LFLAG] & ISIG, ISIG)
            self.assertEqual(int_(mode[CC][VMIN]), 0)
            self.assertEqual(int_(mode[CC][VTIME]), 1)
        self.test_defaults()

    def test_setraw_raises_on_bad_fd(self):
        with open('/dev/null', MODE) as stdin:
            self.assertRaises(termios.error, setraw, stdin)

    def test_setcbreak_raises_on_bad_fd(self):
        with open('/dev/null', MODE) as stdin:
            self.assertRaises(termios.error, setcbreak, stdin)

    def test_rawmode_raises_on_bad_fd(self):
        with open('/dev/null', MODE) as stdin:
            self.assertRaises(termios.error, rawmode(stdin).__enter__)

    def test_cbreakmode_raises_on_bad_fd(self):
        with open('/dev/null', MODE) as stdin:
            self.assertRaises(termios.error, cbreakmode(stdin).__enter__)

    def test_setraw_raises_on_None_fd(self):
        self.assertRaises(TypeError, setraw, None)

    def test_setcbreak_raises_on_None_fd(self):
        self.assertRaises(TypeError, setcbreak, None)

    def test_rawmode_raises_on_None_fd(self):
        self.assertRaises(TypeError, rawmode(None).__enter__)

    def test_cbreakmode_raises_on_None_fd(self):
        self.assertRaises(TypeError, cbreakmode(None).__enter__)

    def test__opentty(self):
        tty = _opentty('/dev/tty', -1)
        self.assertNotEqual(tty, None)
        tty.close()

    def test__opentty_unbuffered(self):
        tty = _opentty('/dev/tty', 0)
        self.assertNotEqual(tty, None)
        tty.close()

    def test__opentty_buffered(self):
        tty = _opentty('/dev/tty', 512)
        self.assertNotEqual(tty, None)
        tty.close()

    def test__opentty_line_buffered(self):
        tty = _opentty('/dev/tty', 1)
        self.assertNotEqual(tty, None)
        tty.close()

    def test_opentty(self):
        with opentty() as tty:
            self.assertNotEqual(tty, None)
            self.assertEqual(sorted(tty.mode), sorted(MODE))

    def test_opentty_accepts_bufsize_argument(self):
        with opentty(-1) as tty:
            self.assertNotEqual(tty, None)

    def test_opentty_returns_None_on_bad_device(self):
        opener = opentty()
        opener.device = '/dev/foobar'
        with opener as tty:
            self.assertEqual(tty, None)

    def test__readyx(self):
        stream = BytesIO(b'\033[24;1R')
        self.assertEqual(_readyx(stream), (24, 1))

    def test__readyx_empty(self):
        stream = BytesIO()
        self.assertEqual(_readyx(stream), (0, 0))

    def test__readyx_too_short(self):
        stream = BytesIO(b'\033[24;1')
        self.assertEqual(_readyx(stream), (0, 0))

    def test__readyx_not_a_number(self):
        stream = BytesIO(b'\033[24;%R')
        self.assertEqual(_readyx(stream), (0, 0))

    def test_getyx(self):
        line, col = getyx()
        self.assertNotEqual(line, 0)
        self.assertNotEqual(col, 0)

    def test_readto_3(self):
        stream = BytesIO(b'123456789')
        self.assertEqual(readto(stream, b'3'), b'123')

    def test_readto_7(self):
        stream = BytesIO(b'123456789')
        self.assertEqual(readto(stream, b'7'), b'1234567')

    def test_readto_4_or_5(self):
        stream = BytesIO(b'123456789')
        self.assertEqual(readto(stream, (b'5', b'4')), b'1234')

    def test_readto_456(self):
        stream = BytesIO(b'123456789')
        self.assertEqual(readto(stream, b'456'), b'123456')

    def test_readto_end_if_bad_stopbyte(self):
        stream = BytesIO(b'123456789')
        self.assertEqual(readto(stream, b'0'), b'123456789')

    def test_readto_end_if_empty_stopbyte(self):
        stream = BytesIO(b'123456789')
        self.assertEqual(readto(stream, b''), b'123456789')

    def test_readto_end_if_empty_stopbyte_in_tuple(self):
        stream = BytesIO(b'123456789')
        self.assertEqual(readto(stream, (b'',)), b'123456789')

    def test_readto_end_if_empty_tuple(self):
        stream = BytesIO(b'123456789')
        self.assertEqual(readto(stream, ()), b'123456789')

    def test_readto_empty_stream(self):
        stream = BytesIO(b'')
        self.assertEqual(readto(stream, b'3'), b'')

    def test_stringio_readto_3(self):
        stream = StringIO('123456789')
        self.assertEqual(readto(stream, '3'), '123')

    def test_stringio_readto_4_or_5(self):
        stream = StringIO('123456789')
        self.assertEqual(readto(stream, ('5', '4')), '1234')

    def test_stringio_readto_456(self):
        stream = StringIO('123456789')
        self.assertEqual(readto(stream, '456'), '123456')

    def test_stringio_readto_end_if_empty_stopchar(self):
        stream = StringIO('123456789')
        self.assertEqual(readto(stream, ''), '123456789')

    def test_stringio_readto_end_if_empty_stopchar_in_tuple(self):
        stream = StringIO('123456789')
        self.assertEqual(readto(stream, ('',)), '123456789')

    def test_stringio_readto_end_if_empty_tuple(self):
        stream = StringIO('123456789')
        self.assertEqual(readto(stream, ()), '123456789')

    def test_stringio_readto_empty_stream(self):
        stream = StringIO('')
        self.assertEqual(readto(stream, '3'), '')

    def test__readcolor(self):
        stream = BytesIO(b'\033]10;rgb:00ff/ff00/0ff0\007')
        self.assertEqual(_readcolor(stream), (255, 65280, 4080))

    def test__readcolor_empty(self):
        stream = BytesIO()
        self.assertEqual(_readcolor(stream), (-1, -1, -1))

    def test__readcolor_too_short(self):
        stream = BytesIO(b'\033]10;rgb:00ff/ff00/0ff0')
        self.assertEqual(_readcolor(stream), (-1, -1, -1))

    def test__readcolor_not_a_hex_number(self):
        stream = BytesIO(b'\033]10;rgb:00ff/ff00/0gg0\007')
        self.assertEqual(_readcolor(stream), (-1, -1, -1))

    def test_luminance_black(self):
        self.assertEqual(luminance((0, 0, 0)), 0)

    def test_luminance_white(self):
        self.assertAlmostEqual(luminance((65535, 65535, 65535)), 65535)

    def test_luminance_oceanblue(self):
        self.assertAlmostEqual(luminance((11103, 26278, 51575)), 27302.756912974924)

    def test_luminance_sandred(self):
        self.assertAlmostEqual(luminance((36524, 13538, 10094)), 22761.062302889117)

    def test_getbgcolor(self):
        rgb = getbgcolor()
        self.assertNotEqual(rgb, (-1, -1, -1))

    def test_getfgcolor(self):
        rgb = getfgcolor()
        self.assertNotEqual(rgb, (-1, -1, -1))

    def test_islightmode(self):
        self.assertNotEqual(islightmode(), None)

    def test_isdarkmode(self):
        self.assertNotEqual(isdarkmode(), None)

    def test_isxterm(self):
        with setterm('xterm-color'):
            self.assertEqual(isxterm(), True)
        with setterm('xterm-256color'):
            self.assertEqual(isxterm(), True)
        with setterm('xterm-'):
            self.assertEqual(isxterm(), True)
        with setterm('xterm'):
            self.assertEqual(isxterm(), True)

    def test_is_not_xterm_(self):
        with setterm('ansi'):
            self.assertEqual(isxterm(), False)
        with setterm('vt100'):
            self.assertEqual(isxterm(), False)



term/__init__.py
"""The term module is intended to replace the tty module."""

# Authors: Steen Lumholt, Stefan H. Holek

import sys
import os
import re

from math import sqrt
from termios import *

__all__ = ["setraw", "setcbreak", "rawmode", "cbreakmode",
           "IFLAG", "OFLAG", "CFLAG", "LFLAG", "ISPEED", "OSPEED", "CC",
           "opentty", "readto", "TIMEOUT",
           "getyx"] # BBB

# Indexes for termios list.
IFLAG = 0
OFLAG = 1
CFLAG = 2
LFLAG = 3
ISPEED = 4
OSPEED = 5
CC = 6

# Open /dev/tty in binary mode by default.
MODE = 'r+b'

# Wait up to 0.2 seconds for a response.
TIMEOUT = 2


def setraw(fd, when=TCSAFLUSH, min=1, time=0):
    """Put the terminal in raw mode.

    Wait until at least `min` bytes or characters have been read.
    If `min` is 0, give up after `time` (in 1/10ths of a second)
    without data becoming available.
    """
    mode = tcgetattr(fd)
    mode[IFLAG] = mode[IFLAG] & ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON)
    mode[OFLAG] = mode[OFLAG] & ~(OPOST)
    mode[CFLAG] = mode[CFLAG] & ~(CSIZE | PARENB)
    mode[CFLAG] = mode[CFLAG] | CS8
    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON | IEXTEN | ISIG)
    mode[CC][VMIN] = min
    mode[CC][VTIME] = time
    tcsetattr(fd, when, mode)


def setcbreak(fd, when=TCSAFLUSH, min=1, time=0):
    """Put the terminal in cbreak mode.

    Wait until at least `min` bytes or characters have been read.
    If `min` is 0, give up after `time` (in 1/10ths of a second)
    without data becoming available.
    """
    mode = tcgetattr(fd)
    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON)
    mode[CC][VMIN] = min
    mode[CC][VTIME] = time
    tcsetattr(fd, when, mode)


class rawmode(object):
    """Context manager to put the terminal in raw mode.

    The current mode is saved and restored on exit.
    """

    def __init__(self, fd, when=TCSAFLUSH, min=1, time=0):
        self.fd = fd
        self.when = when
        self.min = min
        self.time = time
        self.savedmode = None

    def __enter__(self):
        self.savedmode = tcgetattr(self.fd)
        setraw(self.fd, self.when, self.min, self.time)

    def __exit__(self, *ignored):
        tcsetattr(self.fd, TCSAFLUSH, self.savedmode)


class cbreakmode(object):
    """Context manager to put the terminal in cbreak mode.

    The current mode is saved and restored on exit.
    """

    def __init__(self, fd, when=TCSAFLUSH, min=1, time=0):
        self.fd = fd
        self.when = when
        self.min = min
        self.time = time
        self.savedmode = None

    def __enter__(self):
        self.savedmode = tcgetattr(self.fd)
        setcbreak(self.fd, self.when, self.min, self.time)

    def __exit__(self, *ignored):
        tcsetattr(self.fd, TCSAFLUSH, self.savedmode)


def _opentty(device, bufsize, mode=MODE):
    """Open a tty device for reading and writing."""
    fd = os.open(device, os.O_RDWR | os.O_NOCTTY)

    # Line buffering is text mode only
    if bufsize == 1:
        bufsize = -1

    if sys.version_info[0] >= 3:
        # Buffering requires a seekable device
        try:
            os.lseek(fd, 0, os.SEEK_CUR)
        except OSError:
            bufsize = 0
        return open(fd, mode, bufsize)
    else:
        return os.fdopen(fd, mode, bufsize)


class opentty(object):
    """Context manager returning a new rw stream connected to /dev/tty.

    The stream is None if the device cannot be opened.
    The `mode` argument must be 'r+b' (default) or 'r+'.
    """
    device = '/dev/tty'

    def __init__(self, bufsize=-1, mode=MODE):
        self.bufsize = bufsize
        self.mode = mode
        self.tty = None

    def __enter__(self):
        self.tty = None
        try:
            self.tty = _opentty(self.device, self.bufsize, self.mode)
        except EnvironmentError:
            pass
        return self.tty

    def __exit__(self, *ignored):
        if self.tty is not None:
            self.tty.close()


def _readto(stream, stopbyte):
    """Read bytes from stream, up to and including stopbyte.

    Returns an empty bytes object on EOF.
    """
    p = b''
    c = stream.read(1)
    while c:
        p += c
        if c == stopbyte:
            break
        c = stream.read(1)
    return p


def readto(stream, endswith):
    """Read bytes or characters from `stream` until buffer.endswith(`endswith`)
    is true.

    The `endswith` argument may be a single suffix or a tuple of
    suffixes to try.
    Suffixes must be bytes or str depending on the stream.
    Empty suffixes are ignored.
    """
    if not isinstance(endswith, (tuple, list)):
        endswith = (endswith,)

    c = stream.read(1)
    if isinstance(c, str):
        p = ''
        endswith = tuple(x for x in endswith if not isinstance(x, str) or x)
    else:
        p = b''
        endswith = tuple(x for x in endswith if not isinstance(x, bytes) or x)

    while c:
        p += c
        if p.endswith(endswith):
            break
        c = stream.read(1)
    return p


# linecol: printf '\033[6n' -> b'\x1b[24;1R'

def _readyx(stream):
    """Read a CPR response from stream."""
    p = readto(stream, b'R')
    if p:
        m = re.search(b'(\\d+);(\\d+)R$', p)
        if m is not None:
            return int(m.group(1), 10), int(m.group(2), 10)
    return 0, 0


def getyx():
    """Return the cursor position as 1-based (line, col) tuple.

    Line and col are 0 if the device cannot be opened or
    does not support DSR 6.
    """
    with opentty() as tty:
        if tty is not None:
            with cbreakmode(tty, min=0, time=TIMEOUT):
                tty.write(b'\033[6n')
                return _readyx(tty)
    return 0, 0


def getyx_text_mode():
    """Return the cursor position as 1-based (line, col) tuple.

    Line and col are 0 if the terminal does not support
    DSR 6.
    """
    with opentty(mode='r+') as tty:
        if tty is not None:
            with cbreakmode(tty, min=0, time=TIMEOUT):
                tty.write('\033[6n')
                tty.flush()
                p = readto(tty, 'R')
                if p:
                    m = re.search(r'(\d+);(\d+)R$', p)
                    if m is not None:
                        return int(m.group(1)), int(m.group(2))
    return 0, 0


def getyx_stdin_stdout():
    """Return the cursor position as 1-based (line, col) tuple.

    Line and col are 0 if the terminal does not support
    DSR 6.
    """
    if sys.stdin.isatty() and sys.stdout.isatty():
        with cbreakmode(sys.stdin, min=0, time=TIMEOUT):
            sys.stdout.write('\033[6n')
            sys.stdout.flush()
            p = readto(sys.stdin, 'R')
            if p:
                m = re.search(r'(\d+);(\d+)R$', p)
                if m is not None:
                    return int(m.group(1)), int(m.group(2))
    return 0, 0


# fgcolor: printf '\033]10;?\007' -> b'\x1b]10;rgb:0000/0000/0000\x07'
# bgcolor: printf '\033]11;?\007' -> b'\x1b]11;rgb:ffff/ffff/ffff\x07'

def _readcolor(stream):
    """Read an RGB color response from stream."""
    p = readto(stream, b'\007')
    if p:
        m = re.search(b'rgb:([0-9a-fA-F]+)/([0-9a-fA-F]+)/([0-9a-fA-F]+)\007$', p)
        if m is not None:
            return int(m.group(1), 16), int(m.group(2), 16), int(m.group(3), 16)
    return -1, -1, -1


def getfgcolor():
    """Return the terminal foreground color as (r, g, b) tuple.

    All values are -1 if the device cannot be opened or does not
    support OSC 10.
    """
    with opentty() as tty:
        if tty is not None:
            with cbreakmode(tty, min=0, time=TIMEOUT):
                tty.write(b'\033]10;?\007')
                return _readcolor(tty)
    return -1, -1, -1


def getbgcolor():
    """Return the terminal background color as (r, g, b) tuple.

    All values are -1 if the device cannot be opened or does not
    support OSC 11.
    """
    with opentty() as tty:
        if tty is not None:
            with cbreakmode(tty, min=0, time=TIMEOUT):
                tty.write(b'\033]11;?\007')
                return _readcolor(tty)
    return -1, -1, -1


def name():
    """Return the TERM environment variable."""
    return os.environ.get('TERM', '')


def isxterm():
    """Return true if the TERM environment variable starts with 'xterm'."""
    return name().startswith('xterm')


def luminance(rgb):
    """Compute perceived brightness of RGB color tuple."""
    # https://alienryderflex.com/hsp.html
    return sqrt(0.299*rgb[0]**2 + 0.587*rgb[1]**2 + 0.114*rgb[2]**2)


def islightmode():
    """Return true if the background color is lighter than the foreground color.

    May return None if the device cannot be opened or
    does not support OSC 10 & 11.
    """
    bgcolor = getbgcolor()
    if bgcolor[0] >= 0:
        fgcolor = getfgcolor()
        if fgcolor[0] >= 0:
            return luminance(bgcolor) > luminance(fgcolor)


def isdarkmode():
    """Return true if the background color is darker than the foreground color.

    May return None if the device cannot be opened or
    does not support OSC 10 & 11.
    """
    bgcolor = getbgcolor()
    if bgcolor[0] >= 0:
        fgcolor = getfgcolor()
        if fgcolor[0] >= 0:
            return luminance(bgcolor) < luminance(fgcolor)


def getnumcolors():
    """Return the number of colors supported by the terminal.

    The result is 0 if the device cannot be opened
    or has no color support.
    """
    # pylint: disable=import-outside-toplevel
    import curses

    with opentty() as tty:
        if tty is not None:
            try:
                curses.setupterm(name(), tty.fileno())
                colors = curses.tigetnum('colors')
                if colors >= 0:
                    return colors
            except curses.error:
                pass
    return 0


def iscolor():
    """Return true if the terminal has color support."""
    return getnumcolors() > 0


def is256color():
    """Return true if the terminal has >= 256 color support."""
    return getnumcolors() >= 256


# Help autosphinx's autoattribute find the constants...
class term:
    # pylint: disable=too-few-public-methods
    IFLAG = IFLAG
    OFLAG = OFLAG
    CFLAG = CFLAG
    LFLAG = LFLAG
    ISPEED = ISPEED
    OSPEED = OSPEED
    CC = CC
    MODE = MODE
    TIMEOUT = TIMEOUT

